### PR TITLE
[MRG+1] Clarify statistics background

### PR DESCRIPTION
**Pull request by cbrnr**
I'm currently reading the background doc, and although this is an excellent introduction there are some minor tweaks that could make it even better.

I've done everything up to the non-parametric section, which I will tackle in another commit.

### CODE REVIEW

**Comment by cbrnr** [comment]
This section is a bit unclear:

> In the case of a 2-tailed paired t-test against 0 (or between two conditions where you have already subtracted them), exchangeability means that we can flip the signs of our data. Therefore, we can construct the null distribution values by taking random subsets of samples (subjects), flipping the sign of their data, and recording the resulting statistic absolute value. The absolute value of the statistic evaluated on the veridical data can then be compared to this distribution, and the p-value is simply the proportion of null distribution values that were smaller.

The preceding text only introduced the 1-sample t-test. Therefore, I would replace "paired t-test" with "1-sample t-test" since these should be identical. OK?

Furthermore, "taking random subsets of samples (subjects), flipping the sign of their data, and recording the resulting statistic absolute value" should be elaborated a bit more. For example, it sounds like you take some subjects and multiply their data with -1. I guess this is not what is happening. Also, what does "recording the resulting statistic absolute value" exactly mean?

Finally, the note talks about a one-sided test - how is this related to the procedure for the two-sided test explained before?

**Comment by agramfort** [comment]
"Therefore, I would replace "paired t-test" with "1-sample t-test" since these should be identical. OK?"

yes would keep both terms them explaining that it's here equivalent.

"Furthermore, "taking random subsets of samples (subjects), flipping the sign of their data, and recording the resulting statistic absolute value" should be elaborated a bit more. For example, it sounds like you take some subjects and multiply their data with -1. I guess this is not what is happening. Also, what does "recording the resulting statistic absolute value" exactly mean?"

you have contrasted the  2 conditions for a paired test you then want to test if the average significantly deviates from 0. Under the null it's zero mean and symmetric distribution. So swapping sign should not change the observed stats. So you build the non-parametric null distribution by doing these random sign flips. Does it help?

**Comment by cbrnr** [comment]
Thanks @agramfort. The previous section states:

> Let's start with a **1-sample t-test**, which is a standard test for differences in paired sample means.

I think it is OK (and less confusing) to use "1-sample t-test" and not introduce yet another term ("paired t-test") which actually means the same thing.

Please see the changes in the latest commit and let me know if this paragraph is OK.

One thing that is not clear to me: in a one-tailed test, the number of permutations (without the veridical distribution) is `2**N - 1`  (1023 for `N = 10` in our example). However, in the two-tailed test this number is onle `2**(N-1) - 1` (511 in our example) - at least this is what the output of `permutation_t_test` tells us. Why is that?

**Comment by cbrnr** [comment]
BTW, why is CircleCI complaining about a `ModuleNotFoundError: No module named 'PyQt5.sip'`?

File: tutorials/plot_background_statistics.py [hunk 1]
```diff
@@ -178,47 +179,45 @@ def plot_t_p(t, p, title, mcc, axes=None):
...
-# In the case of a 2-tailed paired t-test against 0 (or between two conditions
-# where you have already subtracted them), exchangeability means that we can
-# flip the signs of our data. Therefore, we can construct the
-# **null distribution** values by taking random subsets of samples (subjects),
-# flipping the sign of their data, and recording the resulting statistic
-# absolute value. The absolute value of the statistic evaluated on the
-# veridical data can then be compared to this distribution, and the p-value
-# is simply the proportion of null distribution values that were smaller.
+# In the case of a two-tailed 1-sample t-test (which also tests the difference
+# between two conditions against zero), exchangeability means that we can flip
```
Comment (jona-sassenhagen) [inline tutorials/plot_background_statistics.py]: I'd explicitly write that you could do a two-condition test via a 1-sampe test on their difference. Otherwise it's confusing because people might think of a 2-sample test when you talk of 2 conditions

File: tutorials/plot_background_statistics.py [hunk 2]
```diff
@@ -383,8 +382,8 @@ def plot_t_p(t, p, title, mcc, axes=None):
 # made, as the entire cluster as a whole is used to reject the null.
 # Moreover, because the test statistic concerns the full data, the null
 # hypothesis (and our rejection of it) refers to the structure of the full
-# data. For more information, see also the
-# `excellent FieldTrip cluster interpretation tutorial <ft_cluster>`_.
+# data. For more information, see also the excellent
```
Comment (jona-sassenhagen) [inline tutorials/plot_background_statistics.py]: Is the "excellent" adequate style here? Isn't "comprehensive" or "authoritative" better?

**Comment by cbrnr** [comment]
Just to make sure I understood the non-parametric test with a maximum statistic: when we say "we take the *maximum statistic across all tests* in each permutation to form the null distribution", we mean we record only the maximum across all voxels per permutation, right? So the difference to the regular non-parametric resampling test is that we do not create the null distribution for each voxel individually, but instead we now have only one null distribution derived from all voxels (and we compare each voxel to this single null distribution). Correct?

Furthermore, I removed any mention of "cluster" here, because this should have nothing to do with clusters yet.

**Comment by agramfort** [comment]
yes correct !

by having such a global null distribution you fix the multiple comparison
pb.

File: tutorials/plot_background_statistics.py [hunk 3]
```diff
@@ -246,29 +255,29 @@ def plot_t_p(t, p, title, mcc, axes=None):
...
+fig.show()
 
 ###############################################################################
-# To combat this problem, multiple methods exist. Typically these
-# provide control over either the:
+# To combat this problem, several methods exist. Typically these
+# provide control over either one of the following two measures:
 #
-# 1. `Familywise error rate (FWER) <fwer>`_
+# 1. `Familywise error rate (FWER) <https://en.wikipedia.org/wiki/Family-wise_error_rate>`_
```
Comment (larsoner) [inline tutorials/plot_background_statistics.py]: PEP8 line length, should probably be reverted

https://travis-ci.org/mne-tools/mne-python/jobs/400807122#L3216

File: tutorials/plot_background_statistics.py [hunk 4]
```diff
@@ -178,47 +187,47 @@ def plot_t_p(t, p, title, mcc, axes=None):
...
 # potentially problematic for these data because there are
-# :math:`40 \times 40 = 1600` tests being performed. If we just use
-# a threshold ``p < 0.05`` for all of our tests, we would expect many
-# voxels to be declared significant even if there were no true effect.
+# :math:`40 \cdot 40 = 1600` tests being performed. If we use a threshold
+# ``p < 0.05`` for each individual test, we would expect many voxels to be
+# declared significant even if there were no true effect.
 # In other words, we would make many **type I errors** (adapted from
-# `here <https://en.wikipedia.org/wiki/Type_I_and_type_II_errors>`_):
+# `here <https://en.wikipedia.org/w/index.php?title=Type_I_and_type_II_errors#Table_of_error_types>`_):  # noqa
```
Comment (larsoner) [inline tutorials/plot_background_statistics.py]: you should probably use a link as done in `master`, the `# noqa` gets rendered

https://8408-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_background_statistics.html#multiple-comparisons

File: tutorials/plot_background_statistics.py [hunk 5]
```diff
@@ -383,8 +392,8 @@ def plot_t_p(t, p, title, mcc, axes=None):
 # made, as the entire cluster as a whole is used to reject the null.
 # Moreover, because the test statistic concerns the full data, the null
 # hypothesis (and our rejection of it) refers to the structure of the full
-# data. For more information, see also the
-# `excellent FieldTrip cluster interpretation tutorial <ft_cluster>`_.
+# data. For more information, see also the comprehensive
+# `FieldTrip tutorial <http://www.fieldtriptoolbox.org/faq/how_not_to_interpret_results_from_a_cluster-based_permutation_test>`_.
```
Comment (larsoner) [inline tutorials/plot_background_statistics.py]: revert linking method change

Comment (cbrnr) [reply tutorials/plot_background_statistics.py]: I changed the links because they didn't work in master: https://martinos.org/mne/stable/auto_tutorials/plot_background_statistics.html

I can remove the `# noqa`, but then PEP8 complains.

**Comment by larsoner** [comment]
Maybe the link URLs need to be in the same code block, can you try it that
way?

**Comment by cbrnr** [comment]
Nope, the links still don't work: https://8413-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_background_statistics.html

**Comment by larsoner** [comment]
Fixed the links using
```
`whatever <errors_>`_
```
Some of the URLs have a dangling `#noqa` that get attached but it shouldn't really do any harm. We can fix those later.

**Comment by cbrnr** [comment]
Hm, the appended `#noqa` is a bit ugly - maybe we find a way to break long URLs somehow. We could break it with a `\` I guess.

I'm not done yet, I will work on the second part and try to clarify further. I'll let you know when it's done so you can have a look.

**Comment by larsoner** [comment]
We could add them to links.inc (with include directive at the top) there is no length restriction there

**Comment by cbrnr** [comment]
OK - `links.inc` seems to be the top-level file, because it only includes other `.inc` files (I don't know what `.. compiling links file` does). The links for this tutorial do not fit in any of these files (`known_projects.inc`, `this_project.inc`, and `git_links.inc`). Should we create a new file? Do you have a suggestion for its name and location? I was thinking we could call it `tutorial_links.inc` and put it directly in the `tutorials` folder.

**Comment by agramfort** [comment]
I would put it the root doc folder.

**Comment by cbrnr** [comment]
Of course this didn't work. @larsoner do you know what went wrong?
https://8422-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_background_statistics.html

**Comment by larsoner** [comment]
You should put it in the same place as all other links, add it to `links.inc`, and include `../links.inc` in your tutorial.

**Comment by larsoner** [comment]
(I *think* this is where the file should be while building, so hopefully it works with that path)

**Comment by larsoner** [comment]
I was wrong, you can't link `../links.inc` because it doesn't find the other files properly, doing `../tutorial_links.inc` (as you were probably going to!) should work though -- building locally I see it properly pull in `links.inc` but then fail to find the others because of path confusion

**Comment by larsoner** [comment]
Locally this change works so you should be good to continue with the substantive work now :)

**Comment by cbrnr** [comment]
On to the substantive work, I have a question regarding the steps in clustering:

> In this case, we again do permutations with a maximum statistic, but, under each permutation, we:
> 
> 1. Threshold the computed statistic with some initial threshold value.
> 2. Cluster points that exceed this threshold (with the same sign) based on adjacency.
> 3. Record the size of each cluster (measured, e.g., by a simple voxel count, or by the sum of voxel t-values within the cluster).

How exactly is the maximum statistic used here? It seems like only the threshold is relevant for identifying clusters.

Bonus point: I think it would be illustrative to give a numerical value for the initial threshold, or at least mention how it is determined.

**Comment by larsoner** [comment]
Under each permutation you end up with `N_i` clusters. The "maximum" bit comes in where you take the maximum across those `N_i` clusters. Your null `H0` is formed from these `n_permutation` maximal values.

The threshold used to select points for clustering is a parameter in the function

**Comment by cbrnr** [comment]
> Under each permutation you end up with N_i clusters. The "maximum" bit comes in where you take the maximum across those N_i clusters. Your null H0 is formed from these n_permutation maximal values.

Got it. This should go into the tutorial. I'll work it into the text.

> The threshold used to select points for clustering is a parameter in the function

Yep, just saw that, it is defined at the very top. I'll move it down so that it is really defined right before it is used. Or I guess we can just use its default value, since we're using a t-statistic anyway. What do you prefer?

Hopefully my final question: Why are you performing one-tailed tests in the cluster tests, whereas you did two-tailed tests in all previous tests? I think two-tailed tests make more sense in this generic example.

**Comment by larsoner** [comment]
Let's calculate the threshold to use so we show people how to do it using `scipy.stats.t.ppf`

I think in the toy example they should maybe all be one-tailed tests because they fit the simulation/expectation better. But either way I agree they should all be 1-tailed or all be 2-tailed tests.

**Comment by cbrnr** [comment]
> Let's calculate the threshold to use so we show people how to do it using `scipy.stats.t.ppf`

OK, I'll just mention that in this case the default argument would have the same effect.

> I think in the toy example they should maybe all be one-tailed tests because they fit the simulation/expectation better. But either way I agree they should all be 1-tailed or all be 2-tailed tests.

But you framed everything with 2-tailed tests (the mean is different from 0). I think this is by far the most common case in real-world analyses, so let's switch everything to 2-tailed tests (even though the nature of the peak allows us to formulate a 1-tailed test, usually we don't know the direction of change).

**Comment by cbrnr** [comment]
@larsoner, is this description correct (specifically, I'm unsure about steps 1 and 2)? Also, step 5 was missing, which to me is pretty essential (but again, I'm not sure if I got this right).

For each permutation:
1. Compute the test statistic for each voxel individually.
2. Threshold the test statistic values.
3. Cluster voxels that exceed this threshold (with the same sign) based on adjacency.
4. Record the size of each cluster (measured, e.g., by a simple voxel count, or by the sum of voxel t-values within the cluster).
5. Use the maximum cluster size to build the null distribution.

**Comment by larsoner** [comment]
You only need to "record" / "use" the maximum cluster size in each permutation. So it might be better to change 4 to say "Compute the size of each cluster", then step 5 can be "Use", "Store", or "Retain" the maximum, or something similar. Or you can combine 4 and 5 to use any of these verbs and say "Retain the size of the largest cluster ... to build the null distribution".

**Comment by cbrnr** [comment]
Done. At least for me the tutorial is now even better and easier to understand (hopefully also for others). OK to merge if CIs come back green from my side.

Last comment: I would have expected that clustering gets rid of the spurious blob to the upper right of the real peak. In this example, this is only the case if using hat variance adjustment or TFCE since this small cluster is connected to the large one. This might be due to our example data and/or specific threshold, but it makes getting the main point across more difficult - namely that clustering improves the estimate because it takes the correlation structure into account. For illustrative purposes, setting `rng = np.random.RandomState(2)` might be preferable:

figure_1

WDYT?

**Comment by larsoner** [comment]
Fine by me

**Comment by cbrnr** [comment]
Looks good: https://8483-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_background_statistics.html

**Comment by larsoner** [comment]
@jona-sassenhagen @agramfort feel free to merge if you are happy, I'm still +1 for merge

**Comment by jona-sassenhagen** [comment]
Looks good on first sight, I will try to give it a look later and hopefully merge.

**Comment by jona-sassenhagen** [comment]
Thanks Clemens.

**Comment by agramfort** [comment]
thanks !
